src/main/resources/db/migration/all/20240214170259__add_external_users_to_cas2_notes.sql

```diff
@@ -0,0 +1,23 @@
+BEGIN TRANSACTION;
+-- add external users to notes table
+ALTER TABLE cas_2_application_notes
+    ADD COLUMN created_by_external_user_id UUID;
+-- add fk for external users
+ALTER TABLE cas_2_application_notes
+    ADD CONSTRAINT fk_created_by_external_user_id
+        FOREIGN KEY(created_by_external_user_id)
+        	REFERENCES external_users(id);
+
+-- make nomis users nullable
+ALTER TABLE cas_2_application_notes
+    ALTER COLUMN created_by_nomis_user_id DROP NOT NULL;
+
+-- must have either nomis or external user
+ALTER TABLE cas_2_application_notes
+    ADD CONSTRAINT has_user
```

Comment:
I'd like to be able to test this at the sql level, but I'm not sure how to?

Reply:
I don't think we have an tests for table constraints anywhere else, but I appreciate why we'd want to test this one as it's non trivial. I suspect it's probably not worth the effort as long as we have some integration tests that create an entity with both types of users?

Reply:
@sudhakar-plr [has previously done this by adding a migration job which provides a way to wrap it with tests](https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1252/commits/12f7b648c47b5f79665c9abb75b8322bfae68dc6). Worth checking out!

Reply:
>@sudhakar-plr [has previously done testing mirgations by adding a migration job which provides a way to wrap it with tests](https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1252/commits/12f7b648c47b5f79665c9abb75b8322bfae68dc6). Worth checking out!

@tahb ooh wow that looks really interesting, so it's creating the migration in Kotlin? That seems a much more readable way of doing things than just adding a sql file :) 
It looks like it would be adding quite a lot to this PR but might be worth it in the long run - might timebox it till tomorrow morning to see how far I get with it

Reply:
I've been grappling with adding a Migration Job to the Migration Service as in the PR above, and it would be a fair bit of work for me, and its a slightly different use case in terms of what we're doing to the db. 
@tahb I wonder if we say we have enough confidence for now that the functionality is tested, as we have integration tests for the whole process of saving, and unit tests around the service and Entity transformation? And in the future we'll look at using the Migration Service.